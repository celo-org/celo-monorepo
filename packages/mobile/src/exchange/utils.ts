import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
