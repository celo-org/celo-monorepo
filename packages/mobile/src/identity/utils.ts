import BigNumber from 'bignumber.js'

export function isUserBalanceSufficient(
  userBalance: string | number | null,
  estimatedTxAmount: number | string
) {
  if (!userBalance || new BigNumber(userBalance).isLessThan(new BigNumber(estimatedTxAmount))) {
    return false
  }

  return true
}
