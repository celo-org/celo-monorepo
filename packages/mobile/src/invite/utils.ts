import { sanitizeMessageBase64 } from '@celo/utils/src/attestations'
import URLSearchParamsReal from '@ungap/url-search-params'
import { Platform } from 'react-native'
import RNInstallReferrer from 'react-native-install-referrer'
import Logger from 'src/utils/Logger'

export const createInviteCode = (privateKey: string) => {
  // TODO(Rossy) we need some scheme to encrypt this PK
  // Buffer.from doesn't expect a 0x for hex input
  const privateKeyHex = privateKey.substring(2)
  return Buffer.from(privateKeyHex, 'hex').toString('base64')
}

// exported for testing
export const extractInviteCode = (inviteFieldInput: string) => {
  const sanitizedCode = sanitizeMessageBase64(inviteFieldInput)
  const regex = new RegExp('([0-9A-Za-z/\\+\\-\\_]*=)')
  const matches = sanitizedCode.match(regex)
  if (matches == null || matches.length === 0) {
    return null
  }
  return '0x' + Buffer.from(matches[0], 'base64').toString('hex')
}

// TODO(cmcewen): Consider web3 utils
export const isValidPrivateKey = (hexEncodedPrivateKey: string): boolean => {
  // First two chars are 0x
  if (hexEncodedPrivateKey.length !== 64 + 2) {
    return false
  }
  if (!hexEncodedPrivateKey.startsWith('0x')) {
    return false
  }
  return true
}

export function extractValidInviteCode(inviteFieldInput: string) {
  const inviteCode = extractInviteCode(inviteFieldInput)
  if (inviteCode == null || !isValidPrivateKey(inviteCode)) {
    return null
  } else {
    return inviteCode
  }
}

interface ReferrerData {
  clickTimestamp: string
  installReferrer: string
  installTimestamp: string
}

interface ReferrerDataError {
  message: string
}

export const getValidInviteCodeFromReferrerData = async () => {
  if (Platform.OS === 'android') {
    const referrerData: ReferrerData | ReferrerDataError = await RNInstallReferrer.getReferrer()
    Logger.info(
      'invite/utils/getInviteCodeFromReferrerData',
      'Referrer Data: ' + JSON.stringify(referrerData)
    )
    if (referrerData && referrerData.hasOwnProperty('installReferrer')) {
      const params = new URLSearchParamsReal(
        decodeURIComponent((referrerData as ReferrerData).installReferrer)
      )
      const inviteCode = params.get('invite-code')
      if (inviteCode) {
        const sanitizedCode = inviteCode.replace(' ', '+')
        // Accept invite codes which are either base64 encoded or direct hex keys
        if (isValidPrivateKey(sanitizedCode)) {
          return sanitizedCode
        }
        return extractValidInviteCode(sanitizedCode)
      }
    }
  }
  return null
}
