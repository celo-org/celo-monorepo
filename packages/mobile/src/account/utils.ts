import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
