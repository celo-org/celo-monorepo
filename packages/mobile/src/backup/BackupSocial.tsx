import Button, { BtnTypes } from '@celo/react-components/components/Button'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import { componentStyles } from '@celo/react-components/styles/styles'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { ScrollView, StyleSheet, Switch, Text, View } from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import { connect } from 'react-redux'
import { setSocialBackupCompleted } from 'src/account'
import { showError } from 'src/alert/actions'
import componentWithAnalytics from 'src/analytics/wrapper'
import { ErrorMessages } from 'src/app/ErrorMessages'
import BackupPhraseContainer, {
  BackupPhraseContainerMode,
  BackupPhraseType,
} from 'src/backup/BackupPhraseContainer'
import { getStoredMnemonic, splitMnemonic } from 'src/backup/utils'
import { Namespaces } from 'src/i18n'
import { headerWithBackButton } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import Logger from 'src/utils/Logger'

interface State {
  mnemonic: string
  mnemonicParts: string[]
  isConfirmChecked: boolean
}

interface StateProps {
  socialBackupCompleted: boolean
  language: string | null
}

interface DispatchProps {
  setSocialBackupCompleted: typeof setSocialBackupCompleted
  showError: typeof showError
}

type Props = WithNamespaces & StateProps & DispatchProps

const mapStateToProps = (state: RootState): StateProps => {
  return {
    language: state.app.language,
    socialBackupCompleted: state.account.socialBackupCompleted,
  }
}

class BackupSocial extends React.Component<Props, State> {
  static navigationOptions = () => ({
    ...headerWithBackButton,
  })

  state: State = {
    mnemonic: '',
    mnemonicParts: [],
    isConfirmChecked: false,
  }

  async componentDidMount() {
    await this.retrieveMnemonic()
  }

  retrieveMnemonic = async () => {
    if (this.state.mnemonic) {
      return
    }

    try {
      const mnemonic = await getStoredMnemonic()
      if (!mnemonic) {
        throw new Error('Mnemonic not stored in key store')
      }
      this.setState({ mnemonic, mnemonicParts: splitMnemonic(mnemonic, this.props.language) })
    } catch (e) {
      Logger.error('BackupSocial/retrieveMnemonic', e)
      this.props.showError(ErrorMessages.FAILED_FETCH_MNEMONIC)
    }
  }

  onPressConfirmSwitch = (value: boolean) => {
    this.setState({
      isConfirmChecked: value,
    })
  }

  onPressDone = () => {
    this.props.setSocialBackupCompleted()
    navigate(Screens.BackupComplete)
  }

  render() {
    const { t, socialBackupCompleted } = this.props
    const {
      mnemonicParts: [firstHalf, secondHalf],
      isConfirmChecked,
    } = this.state

    return (
      <SafeAreaView style={styles.container}>
        <ScrollView contentContainerStyle={styles.scrollContainer}>
          <View>
            {!socialBackupCompleted && (
              <>
                <Text style={fontStyles.h1}>{t('setUpSocialBackup')}</Text>
                <Text style={styles.bodyText}>{t('socialBackup.body')}</Text>
              </>
            )}
            {socialBackupCompleted && (
              <>
                <Text style={fontStyles.h1}>{t('socialBackup.yourSafeguards')}</Text>
                <Text style={styles.bodyText}>{t('socialBackupIntro.body')}</Text>
              </>
            )}
            <BackupPhraseContainer
              value={firstHalf}
              showCopy={true}
              mode={BackupPhraseContainerMode.READONLY}
              type={BackupPhraseType.SOCIAL_BACKUP}
            />
            <BackupPhraseContainer
              value={secondHalf}
              showCopy={true}
              mode={BackupPhraseContainerMode.READONLY}
              type={BackupPhraseType.SOCIAL_BACKUP}
              style={componentStyles.marginTop20}
            />
          </View>
          {!socialBackupCompleted && (
            <View style={styles.confirmationSwitchContainer}>
              <Switch
                value={isConfirmChecked}
                onValueChange={this.onPressConfirmSwitch}
                trackColor={switchTrackColors}
                thumbColor={colors.celoGreen}
              />
              <Text style={styles.confirmationSwitchLabel}>{t('socialBackup.confirmation')}</Text>
            </View>
          )}
        </ScrollView>
        {!socialBackupCompleted && (
          <Button
            disabled={!isConfirmChecked}
            onPress={this.onPressDone}
            text={t('global:done')}
            standard={false}
            type={BtnTypes.PRIMARY}
          />
        )}
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
    justifyContent: 'space-between',
  },
  scrollContainer: {
    flexGrow: 1,
    paddingHorizontal: 20,
    paddingBottom: 20,
  },
  bodyText: {
    ...fontStyles.body,
    marginBottom: 20,
  },
  confirmationSwitchContainer: {
    paddingVertical: 20,
    flexDirection: 'row',
  },
  confirmationSwitchLabel: {
    ...fontStyles.body,
    ...fontStyles.semiBold,
    paddingTop: 5,
    paddingLeft: 8,
    paddingRight: 5,
  },
})

const switchTrackColors = {
  false: colors.inactive,
  true: colors.celoGreen,
}

export default componentWithAnalytics(
  connect<StateProps, DispatchProps, {}, RootState>(
    mapStateToProps,
    {
      setSocialBackupCompleted,
      showError,
    }
  )(withNamespaces(Namespaces.backupKeyFlow6)(BackupSocial))
)
