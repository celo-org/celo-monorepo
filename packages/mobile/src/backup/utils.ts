import { sampleSize } from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  const randomWordIndexList = sampleSize([...Array(allWords.length).keys()], numOptions - 1)
  let randomWordIndex: number = 0

  for (let i = 0; i < numOptions; i++) {
    if (i === correctWordPosition) {
      wordOptions.push(correctWord)
      continue
    }

    wordOptions.push(allWords[randomWordIndexList[randomWordIndex]])
    randomWordIndex += 1
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
