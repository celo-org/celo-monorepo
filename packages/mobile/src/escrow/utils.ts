import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
