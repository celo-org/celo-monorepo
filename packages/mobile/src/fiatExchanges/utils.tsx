import { SIMPLEX_URI } from 'src/config'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

export const openMoonpay = (currencyCode: LocalCurrencyCode, currencyToBuy: CURRENCY_ENUM) => {
  navigate(Screens.MoonPay, {
    localAmount: 0,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}
