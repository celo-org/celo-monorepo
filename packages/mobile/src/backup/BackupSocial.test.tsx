import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import BackupSocial from 'src/backup/BackupSocial'
import { createMockStore } from 'test/utils'
import { mockMnemonic } from 'test/values'

jest.mock('react-native-secure-key-store', () => {
  return {
    get: () => {
      return mockMnemonic
    },
  }
})

describe('BackupSocialIntro', () => {
  it('renders correctly when social backup is not complete', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          account: { backupCompleted: true, socialBackupCompleted: false },
        })}
      >
        <BackupSocial />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when social backup is complete', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({ account: { backupCompleted: true, socialBackupCompleted: true } })}
      >
        <BackupSocial />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
