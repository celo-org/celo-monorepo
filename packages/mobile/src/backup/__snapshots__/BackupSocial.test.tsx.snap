// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupSocialIntro renders correctly when social backup is complete 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingBottom": 20,
        "paddingHorizontal": 20,
      }
    }
  >
    <View>
      <View>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Light",
              "fontSize": 22,
              "paddingBottom": 20,
              "textAlign": "center",
            }
          }
        >
          socialBackup.yourSafeguards
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "marginBottom": 20,
            }
          }
        >
          socialBackupIntro.body
        </Text>
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 2,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              socialBackupPhraseHeader
            </Text>
            <View
              accessible={true}
              isTVSelectable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackgroundBorderless",
                  "type": "ThemeAttrAndroid",
                }
              }
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                global:copy
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignContent": "center",
                "backgroundColor": "#EEEEEE",
                "borderRadius": 4,
                "justifyContent": "center",
                "marginTop": 10,
                "padding": 14,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 2,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              socialBackupPhraseHeader
            </Text>
            <View
              accessible={true}
              isTVSelectable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackgroundBorderless",
                  "type": "ThemeAttrAndroid",
                }
              }
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                global:copy
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignContent": "center",
                "backgroundColor": "#EEEEEE",
                "borderRadius": 4,
                "justifyContent": "center",
                "marginTop": 10,
                "padding": 14,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</SafeAreaView>
`;

exports[`BackupSocialIntro renders correctly when social backup is not complete 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingBottom": 20,
        "paddingHorizontal": 20,
      }
    }
  >
    <View>
      <View>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Light",
              "fontSize": 22,
              "paddingBottom": 20,
              "textAlign": "center",
            }
          }
        >
          setUpSocialBackup
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "marginBottom": 20,
            }
          }
        >
          socialBackup.body
        </Text>
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 2,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              socialBackupPhraseHeader
            </Text>
            <View
              accessible={true}
              isTVSelectable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackgroundBorderless",
                  "type": "ThemeAttrAndroid",
                }
              }
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                global:copy
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignContent": "center",
                "backgroundColor": "#EEEEEE",
                "borderRadius": 4,
                "justifyContent": "center",
                "marginTop": 10,
                "padding": 14,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 2,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              socialBackupPhraseHeader
            </Text>
            <View
              accessible={true}
              isTVSelectable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackgroundBorderless",
                  "type": "ThemeAttrAndroid",
                }
              }
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "#42D689",
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                global:copy
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignContent": "center",
                "backgroundColor": "#EEEEEE",
                "borderRadius": 4,
                "justifyContent": "center",
                "marginTop": 10,
                "padding": 14,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "paddingVertical": 20,
          }
        }
      >
        <AndroidSwitch
          accessibilityRole="button"
          enabled={true}
          on={false}
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          thumbTintColor="#42D689"
          trackColorForFalse="#D1D5D8"
          trackColorForTrue="#42D689"
          trackTintColor="#D1D5D8"
        />
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-SemiBold",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingLeft": 8,
              "paddingRight": 5,
              "paddingTop": 5,
            }
          }
        >
          socialBackup.confirmation
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
        },
        null,
        null,
        undefined,
        Object {
          "backgroundColor": "#A3EBC6",
        },
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "type": "ThemeAttrAndroid",
        }
      }
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#A3EBC6",
          },
          Object {
            "height": 50,
          },
          Object {
            "borderWidth": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
              },
              Object {
                "color": "#FFFFFF",
              },
              Object {
                "paddingLeft": 5,
                "paddingRight": 5,
              },
            ]
          }
        >
          global:done
        </Text>
      </View>
    </View>
  </View>
</SafeAreaView>
`;
